p2p/syncer.go

```diff
@@ -294,21 +295,80 @@ func (s *Syncer) syncToTarget(p *Peer, target sunyata.ChainIndex) {
 		// request those transactions
 		for !sc.FullyValidated() {
 			unvalidated := sc.Unvalidated()
-			if len(unvalidated) > 10 {
-				unvalidated = unvalidated[:10]
-			}
-			blocks := getBlocks(unvalidated)
-			if blocks == nil {
-				return
+
+			chunks := len(unvalidated) / BlocksPerRequest
+			if len(unvalidated)%BlocksPerRequest != 0 {
+				chunks++
 			}
-			sc, err = s.cm.AddBlocks(blocks)
-			if errors.Is(err, chain.ErrUnknownIndex) {
-				p.warn(fmt.Errorf("syncLoop: non-attaching blocks: %w", err))
-				return
-			} else if err != nil {
-				s.setErr(fmt.Errorf("syncLoop: couldn't add blocks: %w", err))
-				return
+			unaddedChunks := make([][]sunyata.Block, chunks)
+
+			var chunkMu sync.Mutex
+			var wg sync.WaitGroup
+			wg.Add(chunks)
+			s.cond.L.Unlock()
+			for i := 0; i < chunks; i++ {
+				go func(chunkIndex int) {
```

Comment:
for a mature blockchain, `chunks` could be >10,000, so this loop could spawn thousands of goroutines.
Also, since we're only "bottlenecked" by the number of peers, we can end up with `len(peers)*16` blocks buffered in memory at a time. Usually `len(peers) <= 8`, but it's not unheard of to have as many as 128 peers, meaning over a thousand blocks could be buffered.

I recommend a worker-oriented approach with a circular buffer, along these lines:
```go
// create circular buffer
chunks := make([][]sunyata.Block, 8)

// spawn workers
type req struct {
    // TODO
}
type resp struct {
    // TODO
}
reqChan := make(chan req, len(chunks))
respChan := make(chan resp, len(chunks))
for i := range chunks {
    go func(i int) {
        for req := range reqChan {
            // call getBlocks and send response down respChan
        }
    }(i)
}

// send initial requests
for i := range chunks {
    reqChan <- req {
        // TODO
    }
}

// process responses and send more requests as needed
cur := 0
for !sc.FullyValidated() {
    r := <-respChan
    chunks[r.chunkIndex] = r.blocks

    // drain as much of the buffer as we can
    for chunks[cur] != nil {
        sc, err = s.cm.AddBlocks(chunk)

        chunks[cur] = nil
        cur = (cur + 1) % len(chunks)
    }

    if moreRequestsNeeded {
        reqChan <- req {
            // TODO
        }
    }
}
close(reqChan)
```
You'll need to flesh this out with proper handling of errors, peer selection, etc., but this is a decent outline of the sort of approach that I think we'll want. You can find some examples of worker-oriented algorithms [here](https://github.com/lukechampine/us/blob/kv/renter/renterutil/strategies.go#L660).